src/engraving/libmscore/segment.cpp

```diff
@@ -2706,28 +2706,63 @@ bool Segment::hasAccidentals() const
     return false;
 }
 
-CrossStaffContent Segment::crossStaffContent() const
-{
-    CrossStaffContent crossStaffContent;
-    for (EngravingItem* el : elist()) {
-        if (el && el->isChordRest()) {
-            if (toChordRest(el)->staffMove() == 0 && !toChordRest(el)->isFullMeasureRest()) {
-                crossStaffContent.movedDown = false;
-                crossStaffContent.movedUp = false;
-                break;
+/************************************************************************
+ * computeCrossBeamType
+ * Looks at the chords of this segment and next segment to detect beams
+ * with alternating stem directions (upDown: this chord is up, next chord
+ * is down; downUp: this chord is down, next chord is up). Needed for
+ * correct horizontal spacing of cross-beam situations.
+ * **********************************************************************/
+
+void Segment::computeCrossBeamType(Segment* nextSeg)
+{
+    _crossBeamType.reset();
+    if (!isChordRestType() || !nextSeg || !nextSeg->isChordRestType()) {
+        return;
+    }
+    bool upDown = false;
+    bool downUp = false;
+    for (EngravingItem* e : elist()) {
+        if (!e || !e->isChordRest() || !e->staff()->visible()) {
+            continue;
+        }
+        ChordRest* thisCR = toChordRest(e);
+        if (!thisCR->visible() || thisCR->isFullMeasureRest()) {
+            continue;
+        }
+        if (!thisCR->beam()) {
+            return;
+        }
+        Beam* beam = thisCR->beam();
+        for (EngravingItem* ee : nextSeg->elist()) {
+            if (!ee || !ee->isChordRest() || !ee->staff()->visible()) {
+                continue;
+            }
+            ChordRest* nextCR = toChordRest(ee);
+            if (!nextCR->visible() || nextCR->isFullMeasureRest()) {
+                continue;
+            }
+            if (!nextCR->beam()) {
+                return;
             }
-            if (toChordRest(el)->staffMove() > 0 && toChordRest(el)->beam()) {
-                crossStaffContent.movedDown = true;
+            if (nextCR->beam() != beam) {
+                continue;
+            }
+            if (thisCR->up() == nextCR->up()) {
+                return;
+            }
+            if (thisCR->up() && !nextCR->up()) {
+                upDown = true;
             }
-            if (toChordRest(el)->staffMove() < 0 && toChordRest(el)->beam()) {
-                crossStaffContent.movedUp = true;
+            if (!thisCR->up() && nextCR->up()) {
+                downUp = true;
+            }
+            if (upDown && downUp) {
+                return;
```

Comment:
This basically means that for any CrossBeamType, it will never happen that both upDown and downUp are true at the same time, is that right? (I was thinking, in that case, you could make it an enum with 3 cases as well. However, I'm not sure whether that has any benefits; it looks like it's quite convenient like it is now.)

Reply:
Indeed. If the same segment contains both a down/up and an up/down situation, we shouldn't correct for any of them.